ARROW-10542: [C#][Flight] Add beginning on flight code for net core

Ulimo opened the pull request:
I closed the previous PR and opened this one, since the original one did not use InternalsVisibleTo and required a version bump on the Apache.Arrow project from netstandard1.3 to netstandard1.5.

This adds basic support for both a flight server and a flight client for Net Core.
Sorry for the massive PR, but I wanted to have some basic functionality in place with atleast some tests that use the put and get flow before creating a PR.
I hope that this PR can help create some discussion on how the interfaces should look etc, and if this looks like an accetable interface for flight in net core. It tries to mimic as much as possible the original gRPC net core interface, but mapping the classes from the network protocol to the C# classes.

This implementation uses InternalsVisibleTo, to hinder a bump of .netstandard version from 1.3 to 1.5.
So the Apache.Arrow project still uses netstandard1.3.
All flight code is in a seperate project Apache.Arrow.Flight

This also required changing build version from 2.2 to 3.0

The code does not include:

* Handshake - the reason is that AspNetCore contains features already for authentication/authorization for gRPC. Can be added later ofcourse.
* DoExchange - I feel that more feedback/discussion is required before DoExchange can be implemented.

Note:
**This has been fixed**
Sourcelink did not work when using grpc.tools to have code compilation in the build step. So I had to generate the grpc code manually for sourcelink to work. This means that there are alot of extra code in this PR that are auto generated.

cc: @eerhardt 
Looks like you are the most active for C#, would be great to get some feedback if this is similar to how you would implement it, or if major changes are required I am ofcourse up for that as well.

JamesNK commented on csharp/src/Apache.Arrow.Flight/Apache.Arrow.Flight.csproj:
```diff
@@ -0,0 +1,18 @@
+
+  <ItemGroup>
+    <PackageReference Include="Google.Protobuf" Version="3.14.0" />
+    <PackageReference Include="Grpc" Version="2.33.1" />
```
```suggestion
    <PackageReference Include="Grpc.Tools" Version="2.33.1" PrivateAssets="All" />
```

JamesNK replied:
Only need the tooling, and only at build.

JamesNK replied:
Actually, `Grpc` doesn't bring in `Grpc.Tools`.

Ulimo replied:
I removed the reference to Grpc

JamesNK commented on csharp/src/Apache.Arrow.Flight/Generated/Flight.cs:
```diff
@@ -0,0 +1,3536 @@
+// <auto-generated>
```
Why are you checking generated code in? You should be able to reference Grpc.Tools and generate this file automatically from the proto file.

Ulimo replied:
I had problems in the build process, with sourcelink, I wrote it in the description, sourcelink could not find the auto generated code.

If you know a fix I would greatly appreciate it :)

JamesNK replied:
I haven't heard of this problem before.

Could you provide some more detail about how didn't it work?

Ulimo replied:
here is the build where I saw the errors: https://github.com/apache/arrow/runs/1412215805

Error log extract:
```
2 Documents with errors:
2d1a642b553a64701a2cd2463307ee00b117363d81a8914a91a6e34b6bae0700 sha256 csharp /Users/runner/work/arrow/arrow/csharp/src/Apache.Arrow/obj/Debug/netcoreapp2.1/Format/Flight.cs
https://raw.githubusercontent.com/apache/arrow/afa5d50e94e7f75d8973aa8022502f47b93740ef/csharp/src/Apache.Arrow/obj/Debug/netcoreapp2.1/Format/Flight.cs
error: url failed NotFound: Not Found
36ececbfdcef40763e3f3effdc3fc42392fd80a6fad48bf431395d0e705d7950 sha256 csharp /Users/runner/work/arrow/arrow/csharp/src/Apache.Arrow/obj/Debug/netcoreapp2.1/Format/FlightGrpc.cs
https://raw.githubusercontent.com/apache/arrow/afa5d50e94e7f75d8973aa8022502f47b93740ef/csharp/src/Apache.Arrow/obj/Debug/netcoreapp2.1/Format/FlightGrpc.cs
error: url failed NotFound: Not Found
sourcelink test failed
```

I have not used sourcelink much myself, but it seems like it is linking it to a github file in this case, so I guessed that it is required to be checked in for it to work.

That atleast solved the issue of building (but I know that it is a very ugly fix/workaround), I searched a bit if I can change the build process is any way. But could not find anything yet.

JamesNK replied:
I think `<EmbedUntrackedSources>true</EmbedUntrackedSources>` will fix this.

Ulimo replied:
Ok, I will give it a shot directly, thanks!

Ulimo replied:
That did the trick, thank you! I removed the generated files and added Grpc.Tools instead.

eerhardt commented on csharp/src/Apache.Arrow.Flight/Apache.Arrow.Flight.csproj:
```diff
@@ -0,0 +1,28 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <None Remove="Format\Flight.proto" />
```
This shouldn't be nececessary. Just having the `<Protobuf Include="path-to\Flight.proto" />` should be enough.

Ulimo replied:
Removed

eerhardt commented on csharp/src/Apache.Arrow.Flight/Apache.Arrow.Flight.csproj:
```diff
@@ -0,0 +1,28 @@
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+    <LangVersion>8.0</LangVersion>
+    <EmbedUntrackedSources>true</EmbedUntrackedSources>
```
I think this may make sense to move to the `csharp\Directory.Build.props` file instead.

Ulimo replied:
This was changed in #8702  so it can be tested/approved seperately.

eerhardt commented on csharp/src/Apache.Arrow.Flight/Apache.Arrow.Flight.csproj:
```diff
@@ -0,0 +1,28 @@
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+    <LangVersion>8.0</LangVersion>
```
If it just works - moving this change to the Directory.Build.props would be fine with me. That way we are using the latest C# for all projects.

Ulimo replied:
This was changed in #8702  so it can be tested/approved seperately.

eerhardt commented on csharp/src/Apache.Arrow.Flight/Reader/RecordBatcReaderImplementation.cs:
```diff
@@ -0,0 +1,110 @@
+        {
+            if (!HasReadSchema)
+            {
+                await ReadSchema();
```
Since this is library code, every call to `await` should have a `.ConfigureAwait(false)` on it.

Ulimo replied:
I added ConfigureAwait on (hopefully) all the awaits.

eerhardt commented on csharp/test/Apache.Arrow.Flight.TestWeb/Apache.Arrow.Flight.TestWeb.csproj:
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
```
Let's move all the tests to `netcoreapp3.1`.

Ulimo replied:
This was changed in #8702  so it can be tested/approved seperately.

eerhardt reviewed (commented):
This is great work @Ulimo! Thanks so much for the contribution. This will greatly help the .NET Arrow ecosystem.

Here's some initial feedback on the structure. I will review the code deeper tomorrow as I get more time.

JamesNK commented on csharp/test/Apache.Arrow.Flight.TestWeb/Apache.Arrow.Flight.TestWeb.csproj:
```diff
@@ -0,0 +1,16 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Grpc" Version="2.33.1" />
```
Not needed

```suggestion
```

Ulimo replied:
This is just for the tests where I use some of the extension methods, such as ToListAsync() etc, Is it considered bad practice to use it even in those cases?

Ulimo replied:
Removed it now, thanks!

JamesNK replied:
Grpc brings in some big (50+ meg) NuGet packages. I'd copy the source code to avoid brining them in.

JamesNK commented on csharp/test/Apache.Arrow.Flight.TestWeb/Apache.Arrow.Flight.TestWeb.csproj:
```diff
@@ -0,0 +1,16 @@
+
+  <ItemGroup>
+    <PackageReference Include="Grpc" Version="2.33.1" />
+    <PackageReference Include="Grpc.AspNetCore" Version="2.27.0" />
```
Use latest version
```suggestion
    <PackageReference Include="Grpc.AspNetCore" Version="2.33.1" />
```

Ulimo commented on csharp/src/Apache.Arrow.Flight/Format/Flight.proto:
```diff
@@ -0,0 +1,335 @@
+﻿/*
```
Removed it and added reference instead to the one you posted. Thank you for the feedback!

Ulimo commented on .github/workflows/csharp.yml:
```diff
@@ -38,8 +38,12 @@ jobs:
       fail-fast: false
       matrix:
-        dotnet: [2.2.103]
+        dotnet: ['3.0']
```
I added #8702 which contains net core version upgrade, C# 8.0 and embed untracked sources.

Ulimo commented on .github/workflows/csharp.yml:
```diff
@@ -38,8 +38,12 @@ jobs:
+      - name: Setup .net core 2.2
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: '2.2.103'
```
This was changed in #8702  so it can be tested/approved seperately.

eerhardt commented:
@Ulimo - how much of this is expected to work end-to-end? I've been trying to use it against an example server I found in the repo: https://github.com/apache/arrow/blob/master/python/examples/flight/server.py.

However, I am not able to successfully call anything against the server. Here is my test program:

```C#
        static async Task Main(string[] args)
        {
            GrpcChannel channel = GrpcChannel.ForAddress("http://localhost:5005");
            FlightClient flightClient = new FlightClient(channel);

            var actions = flightClient.ListActions();
            while (await actions.MoveNext(default))
            {
                Console.WriteLine(actions.Current);
            }

            System.Console.WriteLine("Done");
        }
```

And I get this exception:

```
Unhandled exception. Grpc.Core.RpcException: Status(StatusCode="Internal", Detail="Error starting gRPC call. HttpRequestException: An error occurred while sending the request. IOException: The response ended prematurely.", DebugException="System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The response ended prematurely.
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)")
   at Grpc.Net.Client.Internal.HttpContentClientStreamReader`2.MoveNextCore(CancellationToken cancellationToken)
   at Apache.Arrow.Flight.StreamReader`2.MoveNext(CancellationToken cancellationToken) in /Users/eerhardt/git/arrow/csharp/src/Apache.Arrow.Flight/StreamReader.cs:line 46
   at ArrowFlightTest.Program.Main(String[] args) in /Users/eerhardt/DotNetTest/ArrowFlightTest/Program.cs:line 17
   at ArrowFlightTest.Program.<Main>(String[] args)
```

One part that confuses me is the example server says the address is `grpc+tcp://localhost:5005`, but the .NET gRPC client throws with that address saying only `http` and `https` schemes are supported. So that's why I used `http://localhost:5005`.

Ulimo commented:
@eerhardt I tested it against that solution myself, I will check why its not working.
I would love to have a docker container of the java application and use testcontainers so we have unit tests that we are compliant with the java stuff. But that would require adding some more in the build process itself to build a container of the java stuff. Maybe in a future PR?

the dot net grpc stuff only accepts http/https thats true I think a utility class or something to change from grpc+tcp to http/https is required. I also hoped that could come in a future PR.

eerhardt commented on csharp/src/Apache.Arrow.Flight/Apache.Arrow.Flight.csproj:
```diff
@@ -0,0 +1,21 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <Protobuf Include="..\..\..\format\Flight.proto" />
```
```suggestion
    <Protobuf Include="..\..\..\format\Flight.proto" Access="internal" />
```

Let's generate the "Protocol" classes as `internal`, as we don't want public callers to use these types directly.

Ulimo replied:
All the protocol classes are now internal.

eerhardt commented on csharp/src/Apache.Arrow.Flight/Ticket.cs:
```diff
@@ -0,0 +1,60 @@
+
+        public ByteString TicketByteString => _ticket.Ticket_;
+
+        public byte[] TicketBytes => _ticket.Ticket_.ToByteArray();
```
Do we really need all 3 of these as public properties?

One concern I have is that properties are supposed to be cheap, but every time you call `TicketBytes` or `TicketString`, it allocates new memory.

Can we get away with just exposing `TicketByteString` publicaly? And callers can turn it into a `byte[]` or `string` as necessary.

Ulimo replied:
Absolutely, great feedback, thanks 👍

eerhardt replied:
Note, this apply elsewhere where we have the same 3 public properties. Ex. `FlightDescriptor.Command`

Ulimo replied:
Removed the 3 properties from all classes they existed in and they now only expose ByteString.

eerhardt commented on csharp/src/Apache.Arrow.Flight/FlightClient.cs:
```diff
@@ -0,0 +1,90 @@
+        {
+            if(criteria == null)
+            {
+                criteria = new Criteria();
```
We could save on some allocations here by doing something like:

```C#
    public class Criteria
    {
        internal static readonly Criteria Empty = new Criteria();
```

and then here instead of always allocating an empty `Criteria` object each time this is called -

```C#
            if (criteria == null)
            {
                criteria = Criteria.Empty;
```

Ulimo replied:
Fixed.

eerhardt commented on csharp/src/Apache.Arrow.Flight/FlightClient.cs:
```diff
@@ -0,0 +1,90 @@
+
+        public IAsyncStreamReader<ActionType> ListActions(Metadata headers = null)
+        {
+            var response = _client.ListActions(new Empty(), headers);
```
Similar comment for `new Empty()`. Instead make an `internal static readonly Empty Instance = new Empty()` field and use it here.

eerhardt commented on csharp/src/Apache.Arrow.Flight/FlightInfo.cs:
```diff
@@ -0,0 +1,69 @@
+    {
+        private readonly Schema _schema;
+        private readonly FlightDescriptor _flightDescriptor;
+        private readonly IList<FlightEndpoint> _flightEndpoints;
```
We should expose the `FlightDescriptor` and `Schema` publicly so callers can inspect them.

eerhardt replied:
We should also expose `TotalRecords` and `TotalByteLength` properties.

eerhardt commented on csharp/src/Apache.Arrow.Flight/Action.cs:
```diff
@@ -0,0 +1,71 @@
+
+namespace Apache.Arrow.Flight
+{
+    public class Action
```
This is kind of an unfortunate name since it conflicts with `System.Action`. Any time someone tries using this class (or System.Action) and they have both namespaces in their `usings`, they will get:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0104	'Action' is an ambiguous reference between 'Apache.Arrow.Flight.Action' and 'System.Action'	ArrowFlightTest	F:\LinuxShare\ArrowFlight\Program.cs	15	Active
```

Ulimo replied:
@eerhardt I think I will actually prefix all names by Flight if that sounds ok to you as well. A lot of the names are quite generic such as Action, Location, Result, Criteria. There mightbe conflicts with other libraries for the other names as well. So I were thinking that having the same prefix gives it some kind of uniformity.

eerhardt replied:
That makes sense to me.

eerhardt commented on csharp/src/Apache.Arrow.Flight/FlightInfo.cs:
```diff
@@ -0,0 +1,69 @@
+            _flightEndpoints = flightEndpoints;
+        }
+
+        public IEnumerable<FlightEndpoint> Endpoints => _flightEndpoints;
```
This should be `IReadOnlyList<FlightEndpoint>` so people get the Count and index into it.

eerhardt commented on csharp/src/Apache.Arrow.Flight/FlightDescriptor.cs:
```diff
@@ -0,0 +1,105 @@
+            return _flightDescriptor;
+        }
+
+        public bool IsCommand => _flightDescriptor.Type == Protocol.FlightDescriptor.Types.DescriptorType.Cmd;
```
I think we should expose the `DescriptorType` as an enum instead of this `bool IsCommand` property. When callers are checking for a `PATH`, they would write `!IsCommand`. But this is an issue because:

1. If it isn't a Command, then what is it? It's not obvious from reading the code.
2. If there ever was a new kind of descriptor added, all existing code would be broken.

Ulimo replied:
This was mostly taken from the java code, I will change it to a enum instead.

eerhardt commented on csharp/src/Apache.Arrow.Flight/StreamWriter.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace Apache.Arrow.Flight
+{
+    public class StreamWriter<TIn, TOut> : IAsyncStreamWriter<TIn>
```
Does this need to be public? If so, we need to give it a different name, since `StreamWriter` is a common .NET type that we wouldn't want exposed from the Arrow library.

Ulimo replied:
changed to internal

eerhardt reviewed (commented):
This is fantastic work, @Ulimo! I'm really excited to get this in, I've wanted a .NET Arrow Flight library for a while, but haven't had the chance to work on it.

I left a few suggestions I found while playing with the code. One last suggestion I'll make is to go through each class and make sure we only make `public` what actually needs to be public. Once we ship an API it is hard to change/break it. So erroring on the side of hiding something until it is absolutely needed to be public is a good approach.

Ulimo commented:
@eerhardt first of all thank you so much for your support and feedback and sorry for this wall of text. Based on your last feedback that the API is hard to change etc in the future, I really took a long sitting to go through all the classes etc, and found some issues that I really needed to address.

This resulted in a small overhaul on the structure etc. I also went through so the gRPC schema information is actually exposed to the user as well. This also resulted in some rewrites. The feedback you have given has been exactly what I needed to come to the conclusions etc on structure, so I am once again extremely thankful for all the feedback you have given.

This ofcourse increased the scope of the PR a bit again, but I hope it helps give a more real look and feel on how the framework will behave, and if the implemented design actually works.

Here comes information on the work that has been done, and also some checklists to see that all the required information is exposed etc, and why some classes are public.

# Major changes

* All exposed classes start with prefix "Flight".
* New project Apache.Arrow.Flight.AspNetCore, required since generated code is internal, this project references grpc.aspnetcore.server which is netcoreapp3.1 only, and contains extensions only to add flight support for aspnetcore projects.
* All flight client methods now follows gRPC API more closely, allows reading response headers for all calls.
* Exposing application metadata required rewriting a bit of the logic in client and servers, read more in its chapter.
* Moved client specific classes under .Client namespace.
* Moved server specific classes under .Server namespace.
* Internal classes are now under the .Internal namespace.
* Only mutual used classes are now in the .Flight namespace.

# Minor changes

* Removed multi property exposure of bytestring, to only bytestring.
* Exposed TotalBytes and TotalRecords in FlightInfo.
* Moved public abstract reader, writer classes to internal namespace since they cant be used by the user.
* Added FlightServerRecordBatchStreamReader to expose FlightDescriptor only to the server since the client wont use it.
* Added FlightClientRecordBatchStreamReader which just extends FlightRecordBatchStreamReader which is now abstract.
* StreamWriter made internal
* PutResult does not expose ArrowBuffer (was taken from java project), instead it exposes bytestring like the other classes. 
  This was done so the user can choose what type of metadata to return, and also follow more closely the other classes.
* Added public static Empty on PutResult, can be useful for implementations when metadata is not used, and empty put result should be returned.
* Added a new test case to get metadata.
* Added a new test case to put metadata.
* Added a new test case for get schema.
* Added a new test case for DoAction.
* Added a new test case for ListFlights.

Test coverage is now above 80% on the non generated code.

# Application metadata addition

To enable users to write and read application metadata, the class FlightRecordBatchStreamingCall was added that
still follow gRPC look and feel, but exposes FlightRecordBatchStreamReader as response stream instead where the user can
get application metadata.

Application metadata can be read with each record batch, it is implemented as a list since hypothetically metadata can
be sent with the schema aswell. Since the packages sent follow this pattern at the moment:

Schema -> record batch -> record batch -> ... -> done

So the first message could contains metadata as well. When a user gets a new record batch the metadata is cleared similar
to how the java implementation is done.

For writing, this is done by an additional method in FlightRecordBatchStreamWriter which looks like this:
```
public Task WriteAsync(RecordBatch message, ByteString applicationMetadata);
```
# gRPC support
Here is a check on what gRPC methods and properties are exposed and can be read/used by a user of the framework, mostly to check that everything that is required to be exposed are exposed at this time.


## gRPC methods exposed or not exposed to the user
* ListFlights - Exposed
* GetFlightInfo - Exposed
* GetSchema - Exposed
* DoGet - Exposed
* DoPut - Exposed
* DoAction - Exposed
* ListActions -Exposed
* **Handshake - Not exposed**
* **DoExchange - Not exposed**

## gRPC properties exposed or not exposed to the user

### HandshakeRequest
**Handshake is not exposed at all**

* **protocol_version - not exposed**
* **payload - not exposed**

### HandshakeResponse
**Handshake is not exposed at all**

* **protocol_version - not exposed**
* **payload - not exposed**

### BasicAuth
**Basic auth is not exposed at all**

* **username - not exposed**
* **password - not exposed**

### ActionType
* Type - exposed
* Description - exposed

### Criteria
* Expression - exposed

### Action
* Type - exposed
* Body - exposed

### Result
* Body - exposed

### SchemaResult
*Schema result is not mapped to its own type, but returns a schema directly*

* Schema - exposed

### FlightDescriptor
* Type - exposed
* Path - exposed
* Cmd - exposed

### FlightInfo

* Schema - exposed
* FlightDescriptor - exposed
* endpoints - exposed
* TotalRecords - exposed
* TotalBytes - exposed

### FlightEndpoint

* Ticket - exposed
* Locations - exposed

### Location

* Uri - exposed

### Ticket

* Ticket - exposed

### FlightData

Flight data is not exposed as a class, but the data is exposed in getStream, startPut for client, and DoGet and DoPut for server.

* **FlightDescriptor**
  * **client can only send flight descriptor for flight data**
  * **server can only read for flight data**
  * Desc: *The descriptor of the data. This is only relevant when a client is starting a new DoPut stream.*
* **Data header - used to read header of message, ex: schema, record batch message etc, not explicitly exposed and used internally only**
* Application Metadata
  * Client can write metadata through *ClientRecordBatchStreamWriter*
  * Client can read metadata through *FlightRecordBatchStreamReader*
  * Server can write metadata through FlightServerRecordBatchStreamWriter
  * Server can read metadata through *FlightRecordBatchStreamReader*
* **Data body - same as data header, exposure is done through RecordBatch**

### PutResult

* Application metadata - exposed

# Classes public or internal

Here is a list of classes and if they are public or internal, and motivation on why they are public.

* FlightRecordBatchStreamingCall - public, required since it exposes FlightRecordBatchStreamReader which is required to read metadata.
* FlightRecordBatchStreamReader - public, required to allow read on schema, flight descriptor, application metadata
* RecordBatcReaderImplementation - internal
* FlightClientRecordBatchStreamWriter - public, implements CompleteAsync for clients, extends FlightRecordBatchStreamWriter wich allows user to write application metadata.
* FlightDataStream - internal
* FlightRecordBatchDuplexStreamingCall - public, exposes FlightClientRecordBatchStreamWriter which is required to write application metadata.
* FlightRecordBatchStreamWriter - public abstract with private protected constructor, client and server implementations extend this one.
* FlightServerRecordBatchStreamWriter - public, extends FlightRecordBatchStreamWriter which allows user to write application metadata.
* SchemaWriter - internal
* FlightAction - public, allows user to read/write type, and body in Actions.
* FlightActionType - public, allows user to read/write action types with type and description from ListActions.
* FlightClient - public, allows user to call all the different endpoints.
* FlightCriteria - public, exposes Expression that servers can implement to filter result from ListFlights.
* FlightDescriptor - public, exposes DescriptorType, Paths and Cmd for user.
* FlightDescriptorType - public, contains descriptor types.
* FlightEndpoint - public, exposes flight ticket and flight locations.
* FlightInfo - public, exposes flight descriptor, schema, total bytes, total records and endpoints.
* FlightLocation - public, exposes uri to the user.
* FlightMessageSerializer - internal
* FlightPutResult - public, recieved when doing doPut, exposes application metadata.
* FlightResult - public, contains body from DoAction calls.
* FlightServerImplementation - internal, forced internal from putting generated code to internal.
* FlightTicket - public, exposes the bytestring from the ticket to the user.
* IFlightServer - public, interface to implement a flight server.
* StreamReader - internal
* StreamWriter - internal
* FlightIEndpointRouteBuilderExtensions - public, allows user to map the flight endpoint in asp net core.
* FlightIGrpcServerBuilderExtensions - public, allows the user to add their implementation of IFlightServer in a nicer way.

Ulimo commented:
@eerhardt  to be honest after I went through and really thought regarding the calls, I had one final thing I was not too happy with.
The record batch stream in for get and put, I did not change it since compared to moving files around etc this felt a tiny bit bigger.

The API in itself feels quite easy to use, but I think I overspecified it for RecordBatch, and there will be trouble if we add dictionary support etc.
It also does not work well with DoExchange, and do exchange requires another implementation to work good.

The solution I can think of would reduce on ease of use thought (I think?), which would be:

* New classes, FlightData (abstract), FlightDataSchema, FlightDataRecordBatch
* Stream returns FlightData
* internal code still checks flight "correctness" (schema first message etc).

FlightData structure:

```
pubic abstract class FlightData {
  public FlightDescriptor FlightDescriptor { get; }
  public FlightDataType Type { get; } //Contains what type of arrow object it was
  public ByteString ApplicationMetadata { get; } //app metadata can be changed from list in streamreader to correct single in FlightData.
  public abstract T GetValue<T>();
  public abstract void Accept(FlightDataVisitor visitor); // Maybe a visitor pattern?
}

public class FlightDataRecordBatch : FlightData {
  public RecordBatch RecordBatch { get; }
  public GetValue<T>() {
    if(typeof(T) != typeof(RecordBatch)) { throw new Exception(...); }
    return RecordBatch;
  }

  public Accept(...) {...};
}

enum FlightDataType {
  Schema = 1,
  RecordBatch = 2,
  ...
}
```

This stream solution could also have a FlightDataVisitor for the different types also, and I think that it allows
easier addition in the future of new arrow objects. But it of course might be a bit harder to use.
It also follows the gRPC schema much more closely, which makes DoExchange simple to implement.

If not using a visitor, the loop would not be as "pretty" but still functional with easier additions of new types (atleast in my opinion):

```
while(stream.MoveNext()) {
  if(stream.Current.Type == FlightDataType.RecordBatch) {
    stream.Current.GetValue<RecordBatch>();
    // do stuff with record batch
  }
}
```

Would love to get your input, so the flight framework starts with a good foundation.

eerhardt commented on csharp/src/Apache.Arrow/Ipc/ArrowStreamWriter.cs:
```diff
@@ -343,7 +343,7 @@ private protected void WriteRecordBatchInternal(RecordBatch recordBatch)
         }
 
-        private Tuple<ArrowRecordBatchFlatBufferBuilder, VectorOffset> PreparingWritingRecordBatch(RecordBatch recordBatch)
+        private protected Tuple<ArrowRecordBatchFlatBufferBuilder, VectorOffset> PreparingWritingRecordBatch(RecordBatch recordBatch)
```
Is this change necessary? I don't see additional callers of this method.

Ulimo replied:
reverted back

eerhardt commented on csharp/src/Apache.Arrow.Flight/Client/FlightClientRecordBatchStreamWriter.cs:
```diff
@@ -0,0 +1,52 @@
+
+        protected override void Dispose(bool disposing)
+        {
+            CompleteAsync().Wait();
```
Is this necessary? Having sync-over-async is typically not good.

What are the ramifications if someone doesn't call CompleteAsync() before disposing? Should we just `throw` in that case?

Ulimo replied:
Added a throw

eerhardt commented on csharp/src/Apache.Arrow.Flight/Client/FlightRecordBatchDuplexStreamingCall.cs:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        //
+        // Summary:
```
Can these be `xml` comments?

eerhardt commented on csharp/src/Apache.Arrow.Flight/Client/FlightRecordBatchDuplexStreamingCall.cs:
```diff
@@ -0,0 +1,92 @@
+
+        //
+        // Summary:
+        //     Gets the call status if the call has already finished. Throws InvalidOperationException
```
How do callers know when the call has finished? Do they `await` on the `ResponseHeadersAsync`?

Ulimo replied:
The caller ends the stream, so they call CompleteAsync on the request stream.

eerhardt commented on csharp/src/Apache.Arrow.Flight/Internal/FlightRecordBatchStreamReader.cs:
```diff
@@ -0,0 +1,65 @@
+    /// Use MoveNext() and Current to iterate over the batches.
+    /// There are also gRPC helper functions such as ToListAsync() etc.
+    /// </summary>
+    public abstract class FlightRecordBatchStreamReader : IAsyncStreamReader<RecordBatch>
```
Can this class also implement `IAsyncEnumerable<RecordBatch>`? Then callers can just directly `await foreach` it, without having to call `ReadAllAsync()`.

My understanding of why `IAsyncStreamReader<T>` doesn't implement `IAsyncEnumerable<T>` is because the Grpc interface came first, and it couldn't be changed without a breaking change. @JamesNK - is this the case?

Ulimo replied:
Implemented IAsyncEnumerable

eerhardt commented on csharp/src/Apache.Arrow.Flight/Internal/FlightRecordBatchStreamReader.cs:
```diff
@@ -0,0 +1,65 @@
+using Google.Protobuf;
+using Grpc.Core;
+
+namespace Apache.Arrow.Flight.Internal
```
Can we move this to the root namespace? I don't really like public types in an `Internal` namespace. It feels like an anti-pattern to me.

Ulimo replied:
moved

eerhardt commented on csharp/src/Apache.Arrow.Flight/Internal/FlightMessageSerializer.cs:
```diff
@@ -0,0 +1,57 @@
+            }
+
+            ByteBuffer schemaBuffer = ArrowReaderImplementation.CreateByteBuffer(buffer.Slice(bufferPosition));
+            var Schema = MessageSerializer.GetSchema(ArrowReaderImplementation.ReadMessage<Flatbuf.Schema>(schemaBuffer));
+            return Schema;
+        }
+
+        internal static Schema DecodeSchema(ByteBuffer schemaBuffer)
+        {
+            var Schema = MessageSerializer.GetSchema(ArrowReaderImplementation.ReadMessage<Flatbuf.Schema>(schemaBuffer));
+            return Schema;
```
(nit) variable names should be camelCased. Or just skip the variable altogether.

```suggestion
            return MessageSerializer.GetSchema(ArrowReaderImplementation.ReadMessage<Flatbuf.Schema>(schemaBuffer));
        }

        internal static Schema DecodeSchema(ByteBuffer schemaBuffer)
        {
            return MessageSerializer.GetSchema(ArrowReaderImplementation.ReadMessage<Flatbuf.Schema>(schemaBuffer));
```

eerhardt reviewed (commented):
Looking good. I dug a bit deeper into the code this time. I think this is close to merging. Just a few more comments.

Ulimo commented on csharp/src/Apache.Arrow.Flight/FlightDescriptor.cs:
```diff
@@ -0,0 +1,102 @@
+
+        public IEnumerable<string> Paths => _flightDescriptor.Path;
+
+        public ByteString Cmd => _flightDescriptor.Cmd;
```
Had to rename the static create methods to CreatePathDescriptor and CreateCommandDescriptor since Command would collide with it.

eerhardt commented:
>  but I think I overspecified it for RecordBatch, and there will be trouble if we add dictionary support etc.

I think this is OK, for now. This is just the initial impementation. The Flight API won't necessarily be "stable" in its first release - even in the C++/Python APIs it isn't stable yet, as far as I understand. There are probably many more places that will need to change to support dictionaries in the future.
Regarding what I said last week about being hard to change later - that was mostly around the concept that it is easier to hide things now, and expose them later if needed than it is to expose things now and hide them later.

eerhardt commented:
Do you think you can fill out this table with the current support?

https://github.com/apache/arrow/blob/master/docs/source/status.rst#flight-rpc

eerhardt commented on csharp/src/Apache.Arrow.Flight/Server/IFlightServer.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Apache.Arrow.Flight.Server
+{
+    public interface IFlightServer
```
What do you think about changing this to an `abstract class` instead, where all the base `virtual` methods just `throw new NotSupportedException()`?

The advantage is that:

1. If a service doesn't want to support something, they don't need to implement the method.
2. (more importantly) If/when new methods get added, we can add them to the base class without a binary breaking change.

Ulimo replied:
Good suggestion, changed to an abstract class and virtual methods.

eerhardt reviewed (approved):
This is great @Ulimo! Thanks for all the awesome work here.

I'll merge this today unless I hear other feedback.

Ulimo commented:
@eerhardt I just have one question, getting this out as a preview nuget package soonish, is that possible?

eerhardt commented:
I believe the next round of releases is scheduled for sometime in January. (See [this recent dev-list thread](https://lists.apache.org/thread.html/r913292d53736affaff5933c1416fedbb193c4ff68ced4bcf1d02e7ec%40%3Cdev.arrow.apache.org%3E))
Would you need something earlier than that? If so, one option would be to build the nuget package yourself (`dotnet pack` in the csharp folder), and push it to a private feed you can use until an official one get pushed to nuget.org.

Potential options for private feeds are Azure DevOps or myget.org.

eerhardt commented:
FYI - @Ulimo, the `3.0.0` nuget packages have been released. You can see them at:

https://www.nuget.org/packages/Apache.Arrow.Flight/
https://www.nuget.org/packages/Apache.Arrow.Flight.AspNetCore/
